Current core formal spec for WebAssembly Exception Handling (#121)
WebAssembly/exception-handling: document/core/exec/runtime.rst

   \production{(reference)} & \reff &::=&
     \REFNULL~t \\&&|&
     \REFFUNCADDR~\funcaddr \\&&|&
     \REFEXTERNADDR~\externaddr \\&&|&
     \EXNREFADDR~\eventaddr~\val^\ast \\
   \production{(value)} & \val &::=&
     \num ~|~ \reff \\
   \end{array}

The administrative value |EXNREFADDR| is an exception reference of type |EXNREF|, representing a thrown exception on the stack, identified by the exception's :ref:`event address <syntax-eventaddr>`, and carrying the exception event's arguments.
If it is uncaught by a |CATCHN| block, then the embedder defines how to handle it. It is an abstract value, and it is an implementation detail how and where its argument values are stored. It is separate from any linear memory.

> 

Author: Done

It is an invariant of the semantics that the length of the byte vector, divided by page size, never exceeds the maximum size of :math:`\memtype`, if present.

.. index:: ! event instance, event, exception attribute, exception tag
   pair: abstract syntax; event instance
   pair: event; instance
.. _syntax-eventinst:

Event Instances
~~~~~~~~~~~~~~~

An *event instance* is the runtime representation of an :ref:`event <syntax-event>`.
It records the :ref:`attribute <syntax-attribute>` and :ref:`function type <syntax-functype>` of its :ref:`type <syntax-eventtype>`.

.. math::
   \begin{array}{llll}
   \production{(event instance)} & \eventinst &::=&
     \{ \EVIATTRIBUTE~\EXCEPTION, \EVITYPE~\functype \} \\
   \end{array}

Because :ref:`events <syntax-event>` :ref:`have type <valid-event>` |eventtype|, it is an invariant of the semantics that the function type :math:`\functype` has void result type.

.. note::
   In the current version of WebAssembly, events may only be exceptions. This may change in future versions.

>    The event address of an event instance is also called an exception tag.

Author: I'm changing this to emphasis, so `*exception tag*`  or did you mean something else there?

>    The event address of an event instance is also called an exception tag.

Author: Changed to `*exception tag*`

~~~~~~~~~~~~~~~~~~~

An *exception instance* is the runtime representation of an :ref:`exception <syntax-exn>` definition.

> It records type :ref:`type <syntax-exntype>` of the exception.

Reviewer: ```suggestion
It records the type :ref:`type <syntax-exntype>` of the exception.
```

Author: I'll change that to 

> It records the :ref:`type <syntax-exntype>` of the exception.

..................

Like labels, exception handlers carry the return arity :math:`n` of the

> respective try block, and their associated branch *target*, which is

Reviewer: ```suggestion
respective |TRY| block, and their associated branch *target*, which is
```

   \end{array}

.. index:: ! administrative instructions, function, function instance, function address, label, frame, instruction, trap, call, memory, memory instance, table, table instance, element, data, segment, event, event instance, event address, exception, reftype
   pair:: abstract syntax; administrative instruction
.. _syntax-trap:
.. _syntax-reffuncaddr:
.. _syntax-invoke:

> .. _syntax-exnrefaddr:

Author: Done, everywhere.

     \REFFUNCADDR~\funcaddr \\ &&|&
     \REFEXTERNADDR~\externaddr \\ &&|&
     \INVOKE~\funcaddr \\ &&|&
     \EXNREFADDR~\eventaddr~val^\ast \\ &&|&
     \THROWADDR~\eventaddr \\ &&|&

>      \CATCHN_n~\{\instr_2^\ast\}~\instr_1^\ast~\END \\ &&|&
>      \LABEL_n\{\instr^\ast\}~\instr^\ast~\END \\ &&|&
>      \FRAME_n\{\frame\}~\instr^\ast~\END \\

Reviewer: ```suggestion
     \LABEL_n\{\instr^\ast\}~\instr^\ast~\END \\ &&|&
     \CATCHN_n\{\instr^\ast\}~\instr^\ast~\END \\ &&|&
     \FRAME_n\{\frame\}~\instr^\ast~\END \\
```

   \end{array}

The |TRAP| instruction represents the occurrence of a trap.
Traps are bubbled up through nested instruction sequences, ultimately reducing the entire program to a single |TRAP| instruction, signalling abrupt termination.

The |REFFUNCADDR| instruction represents :ref:`function reference values <syntax-ref.func>`. Similarly, |REFEXTERNADDR| represents :ref:`external references <syntax-ref.extern>`,
, and |EXNREFADDR| represents :ref:`exception reference values <syntax-exnrefaddr>` of thrown exception events.

The |INVOKE| instruction represents the imminent invocation of a :ref:`function instance <syntax-funcinst>`, identified by its :ref:`address <syntax-funcaddr>`.
It unifies the handling of different forms of calls.

The |THROWADDR| instruction represents the imminent throw of an :ref:`exception reference value <syntax-exnrefaddr>` based on an :ref:`event instance <syntax-eventinst>`, identified by its :ref:`address <syntax-eventaddr>`.
The values it will consume to create the exception depend on the event's :ref:`event type <syntax-eventtype>`.
It unifies the throwing of different forms of events.

> The |LABEL|, |FRAME|, and |CATCHN| instructions model :ref:`labels <syntax-label>`, :ref:`frames <syntax-frame>`, and :ref:`try-catch blocks <syntax-try>` respectively, :ref:`"on the stack" <exec-notation>`.

Reviewer: ```suggestion
The |LABEL|, |FRAME|, and |CATCHN| instructions model :ref:`labels <syntax-label>`, :ref:`frames <syntax-frame>`, and active :ref:`exception handlers <syntax-try>`, respectively, :ref:`"on the stack" <exec-notation>`.
```

   The selected label is identified through the :ref:`label index <syntax-labelidx>` :math:`l`, which corresponds to the number of surrounding |LABEL| instructions that must be hopped over -- which is exactly the count encoded in the index of a block context.

.. index:: ! throw context, event, exception, throw address, catch block
.. _syntax-ctxt-throw:

Throw Contexts
..............

In order to specify the reduction of :ref:`try-catch blocks <syntax-try>`
with the help of the administrative instructions |THROWADDR|, |EXNREFADDR|, and |CATCHN|,
the following syntax of *throw contexts* is defined, as well as associated structural rules:

.. math::
   \begin{array}{llll}
   \production{(throw contexts)} & \XT &::=&
     \val^\ast~[\_]~\instr^\ast \\ &&|&
     \LABEL_n\{\instr^\ast\}~\XT~\END \\ &&|&
     \CATCHN_n\{\instr^\ast\}~\XT~\END \\ &&|&
     \FRAME_n\{F\}~\XT~\END \\
   \end{array}

.. math::
   \begin{array}{rcl}

>    S;~F;~\CATCHN_m~\{\instr^\ast\}~\val^m~\END &\stepto& S;~F;~\val^m \\

Reviewer: These reduction rules seem out of place here. I think they all belong into instructions.rst, probably into an auxiliary section on exception handlers there, analogous to the ones for blocks.

Author: Done.

>    S;~F;~\CATCHN_m~\{\instr^\ast\}~\val^m~\END &\stepto& S;~F;~\val^m \\
>    S;~F;~\CATCHN_m~\{\instr^\ast\}~\XT[\val^n~[\_]~(\THROWADDR~a)]~\END &\stepto&
>       S;~F;~\LABEL_m~\{\}~(\EXNREFADDR~a~\val^n)~{\instr}^\ast~\END \\

Reviewer: ```suggestion
   S;~F;~\CATCHN_m\{\instr^\ast\}~\val^m~\END &\stepto& S;~F;~\val^m \\
   S;~F;~\CATCHN_m\{\instr^\ast\}~\XT[\val^n~[\_]~(\THROWADDR~a)]~\END &\stepto&
      S;~F;~\LABEL_m\{\}~(\EXNREFADDR~a~\val^n)~{\instr}^\ast~\END \\
```

   && \hspace{-12ex} (\iff S.\SEVENTS[a]=\{\EVATTRIBUTE~\EXCEPTION, \EVTYPE~[t^n]\to[]\}) \\
   \end{array}

> Throw contexts allow to potentially capture the throw of an exception event by surrounding catch blocks.

Reviewer: ```suggestion
Throw contexts allow matching the program context around a throw instruction up to the nearest enclosing :math:`\CATCHN` handler, thereby selecting the exception handler responsible for an exception.
```

.. note::
   For example, catching a simple :ref:`throw <exec-throw>` in a :ref:`try-catch block <exec-try>` would be as follows.

   Let :math:`x` be the :ref:`event index <syntax-eventidx>` of an event in :math:`F.\AMODULE`, such that

   .. math::
      F.\AMODULE.\MIEVENTS[x] = \{ \EVIATTRIBUTE~\EXCEPTION, \EVITYPE~[t^n]\to[]\} = S.\SEVENTS[a]

>    Let :math:`\expand_F(\X{bt}) = [t^n] \to [\EXNREF]`, and consider arbitrary :ref:`values <syntax-val>` :math:`(val : t)^n`.

Reviewer: This seems tangential to the example, maybe remove?

Author: I added the bt expansion to emphasize that this instruction will return the exception reference as a value, not throw it. Then I guess I got carried away and added all the details.  Since my initial point is also spelled out in the end of the example, I will remove these lines as you suggest.

Author: Done.

.. index:: ! throw context, exception, throw address, catch block
.. _syntax-ctxt-throw:

Throw Contexts
..............

In order to specify the reduction of :ref:`try-catch blocks <syntax-try>`
with the help of the administrative instructions |THROWADDR|, |REFEXNADDR|, and |CATCHN|,
the following syntax of *throw contexts* is defined, as well as associated structural rules:

.. math::
   \begin{array}{llll}
   \production{(throw contexts)} & \XT &::=&
     \val^\ast~[\_]~\instr^\ast \\ &&|&
     \LABEL_n\{\instr^\ast\}~\XT~\END \\ &&|&
     \CATCHN_n\{\instr^\ast\}~\XT~\END \\ &&|&
     \FRAME_n\{F\}~\XT~\END \\
   \end{array}

Throw contexts allow matching the program context around a throw instruction up to the nearest enclosing :math:`\CATCHN` handler, thereby selecting the exception handler responsible for an exception.

.. note::
   For example, catching a simple :ref:`throw <exec-throw>` in a :ref:`try-catch block <exec-try>` would be as follows.

   .. math::
      \begin{array}{ll}
      & \hspace{-5ex} S;~F;~\val^n~(\TRY~\X{bt}~\THROW~x~\CATCH~\RETURN~\END) \\
      \stepto & S;~F;~\CATCHN_1\{\RETURN\}~(\LABEL_1 \{\}~\val^n~\THROWADDR~a~\END)~\END \\
      \end{array}

>    :ref:`Handling <exec-handle-exn>` the thrown exception address :math:`a` in the throw context

Reviewer: ```suggestion
   :ref:`Handling <exec-handler-throwaddr>` the thrown exception address :math:`a` in the throw context
```

   When a throw occurs, execution halts until that throw is the continuation of a throw context in a catching try block.

   In this particular case, the exception reference is returned normally, as opposed to being thrown as the result of a plain

>    :math:`\THROW~x` without a catching try block.

Reviewer: ```suggestion
   :math:`\THROW~x` without a catching |TRY| block.
```
